File: 54. Spiral Matrix/54.go
package leetcode

func spiralOrder(matrix [][]int) []int {
	if len(matrix) == 0 || len(matrix[0]) == 0{
        return []int{}
	}
	
	res := make([]int, 0, len(matrix)*len(matrix[0]))

	for len(matrix) != 0 && len(matrix[0]) != 0{
		for _,v := range matrix[0]{
			res = append(res, v)
		}
		if len(matrix) == 0 || len(matrix[0]) == 0{
			break
		}
		matrix = matrix[1:]

		tmp := make([][]int, 0, len(matrix))
		for _,v := range matrix{
			res = append(res, v[len(matrix[0])-1])
			tmp = append(tmp, v[:len(matrix[0])-1])
		}
		if len(matrix) == 0 || len(matrix[0]) == 0{
			break
		}
		matrix = tmp

		for i:=len(matrix[0])-1; i>-1; i--{
			res = append(res, matrix[len(matrix)-1][i])
		}
		if len(matrix) == 0 || len(matrix[0]) == 0{
			break
		}
		matrix = matrix[:len(matrix)-1]

		
		for i:=len(matrix)-1; i>-1; i--{
			res = append(res, matrix[i][0])
		}
		if len(matrix) == 0 || len(matrix[0]) == 0{
			break
		}
		tmp1 := make([][]int, 0, len(matrix))
		for _,v := range matrix{
			tmp1 = append(tmp1, v[1:])
		}
		matrix = tmp1
	}
	
	return res
}

func spiralOrder1(matrix [][]int) []int {
	if len(matrix) == 0 || len(matrix[0]) == 0 {
		return []int{}
	}

	// right->down->left->up->right
	up, down, left, right := 0, len(matrix)-1, 0, len(matrix[0])-1
	i, j := 0, 0
	move := 0
	ret := make([]int, 0)
	processLoop:
	for {
		ret = append(ret, matrix[i][j])
		switch move%4 {
		case 0: // →
			if j == right {
				if i < down {
					i++
					move++
					up++
					continue processLoop
				} else {
					break processLoop
				}
			}
			j++

		case 1: // ↓
			if i == down {
				if j > left {
					j--
					move++
					right--
					continue processLoop
				} else {
					break processLoop
				}
			}
			i++

		case 2: // ←
			if j == left {
				if i > up {
					i--
					move++
					down--
					continue processLoop
				} else {
					break processLoop
				}
			}
			j--

		case 3: // ↑
			if i == up {
				if j < right {
					j++
					move++
					left++
					continue processLoop
				} else {
					break processLoop
				}
			}
			i--
		}
	}
	return ret
}

File: 54. Spiral Matrix/_54.py
class Solution:
    def spiralOrder(self, matrix):
        res = list()
        while len(matrix) != 0 and len(matrix[0]) != 0:
            for i in matrix[0]:
                res.append(i)
            if len(matrix) == 0 or len(matrix[0]) == 0:
                break
            matrix = matrix[1:]

            for i in matrix:
                res.append(i[-1])
            if len(matrix) == 0 or len(matrix[0]) == 0:
                break
            matrix = [i[:-1] for i in matrix]

            for i in matrix[-1][::-1]:
                res.append(i)
            if len(matrix) == 0 or len(matrix[0]) == 0:
                break
            matrix = matrix[:-1]

            for i in matrix[::-1]:
                res.append(i[0])
            if len(matrix) == 0 or len(matrix[0]) == 0:
                break
            matrix = [i[1:] for i in matrix]
        return res

    # 这代码太神了，不太好懂，但思路很好
    def spiralOrder1(self, matrix):
        return matrix and [*matrix.pop(0)] + self.spiralOrder([*zip(*matrix)][::-1])

# Here's how the matrix changes by always extracting the first row and rotating the remaining matrix counter-clockwise:

#     |1 2 3|      |6 9|      |8 7|      |4|  =>  |5|  =>  ||
#     |4 5 6|  =>  |5 8|  =>  |5 4|  =>  |5|
#     |7 8 9|      |4 7|
# Now look at the first rows we extracted:

#     |1 2 3|      |6 9|      |8 7|      |4|      |5|
# Those concatenated are the desired result.

    def spiralOrder2(self, matrix):
        res = []
        if not matrix:
            return []
        i,j,di,dj = 0,0,0,1
        m, n = len(matrix),len(matrix[0])
        for _ in range(m * n):
            res.append(matrix[i][j])
            matrix[i][j] = ''
            if matrix[(i+di)%m][(j+dj)%n] == '':
                di, dj = dj, -di
            i += di
            j += dj
        return res

a = Solution()
print(a.spiralOrder([[1, 2, 3, 4],[5, 6, 7, 8],[9,10,11,12]]))